src/connector/src/source/google_pubsub/source/reader.rs

```diff
@@ -0,0 +1,82 @@
+use std::error::Error;
+
+
+use anyhow::{anyhow, Ok, Result};
+use async_trait::async_trait;
+use google_cloud_pubsub::client::Client;
+use google_cloud_pubsub::subscription::{Subscription, SubscriptionConfig};
+
+use crate::source::google_pubsub::PubsubProperties;
+use crate::source::{Column, ConnectorState, SourceMessage, SplitReader};
+
+const PUBSUB_MAX_FETCH_MESSAGES: usize = 1024;
+
+pub struct PubsubSplitReader {
+    subscription: Subscription,
+}
+
+impl PubsubSplitReader {}
+
+#[async_trait]
+impl SplitReader for PubsubSplitReader {
+    type Properties = PubsubProperties;
+
+    async fn new(
+        properties: PubsubProperties,
+        _state: ConnectorState,
+        _columns: Option<Vec<Column>>,
+    ) -> Result<Self> {
+        let client = Client::default().await.map_err(|e| anyhow!(e))?;
+
+        let topic = client.topic(&properties.topic);
+        let subscription = client.subscription(&properties.subscription);
+
+        // ! Perhaps the split enumerator should be responsible for creating subscriptions.
+        if !subscription
+            .exists(None, None)
+            .await
+            .map_err(|e| anyhow!(e))?
+        {
+            subscription
+                .create(
+                    topic.fully_qualified_name(),
+                    SubscriptionConfig::default(),
+                    None,
+                    None,
+                )
+                .await
+                .map_err(to_anyhow)?;
+        }
+
+        Ok(Self { subscription })
+    }
+
+    async fn next(&mut self) -> Result<Option<Vec<SourceMessage>>> {
+        let next_batch = self
+            .subscription
+            .pull(PUBSUB_MAX_FETCH_MESSAGES as i32, None, None)
+            .await
+            .map_err(to_anyhow)?;
+
+        if next_batch.is_empty() {
```

Comment:
maybe we should wait until the next message. the reader should never return unless it reaches the stop offset.

Reply:
[`subscription.pull`](https://docs.rs/google-cloud-pubsub/latest/google_cloud_pubsub/subscription/struct.Subscription.html#method.pull) blocks until at least one message is available. The check might be unnecessary here, but should we remove it?

Reply:
yes, I think we can remove it